src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -0,0 +1,114 @@
+package jdk.internal.foreign;
+
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.ValueLayout;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+import static java.lang.foreign.ValueLayout.JAVA_BYTE;
+import static java.lang.foreign.ValueLayout.JAVA_SHORT;
+
+/**
+ * Miscellaneous functions to read and write strings, in various charsets.
+ */
+public class StringSupport {
+
+    @ForceInline
+    public static String read(MemorySegment segment, long offset, Charset charset) {
+        return switch (CharsetKind.of(charset)) {
+            case SINGLE_BYTE -> readFast_byte(segment, offset, charset);
+            case DOUBLE_BYTE -> readFast_short(segment, offset, charset);
+            default -> throw new UnsupportedOperationException("Unsupported charset: " + charset);
+        };
+    }
+
+    @ForceInline
+    public static void write(MemorySegment segment, long offset, Charset charset, String string) {
+        switch (CharsetKind.of(charset)) {
+            case SINGLE_BYTE -> writeFast_byte(segment, offset, charset, string);
+            case DOUBLE_BYTE -> writeFast_short(segment, offset, charset, string);
+            default -> throw new UnsupportedOperationException("Unsupported charset: " + charset);
+        }
+    }
+
+    @ForceInline
+    private static String readFast_byte(MemorySegment segment, long offset, Charset charset) {
+        long len = strlen_byte(segment, offset);
+        byte[] bytes = new byte[(int)len];
+        MemorySegment.copy(segment, ValueLayout.JAVA_BYTE, offset, bytes, 0, (int)len);
+        return new String(bytes, charset);
+    }
+
+    @ForceInline
+    private static void writeFast_byte(MemorySegment segment, long offset, Charset charset, String string) {
+        byte[] bytes = string.getBytes(charset);
+        MemorySegment.copy(bytes, 0, segment, ValueLayout.JAVA_BYTE, offset, bytes.length);
+        segment.set(ValueLayout.JAVA_BYTE, offset + bytes.length, (byte)0);
+    }
+
+    @ForceInline
+    private static String readFast_short(MemorySegment segment, long offset, Charset charset) {
+        long len = strlen_short(segment, offset);
+        byte[] bytes = new byte[(int)len];
+        MemorySegment.copy(segment, ValueLayout.JAVA_BYTE, offset, bytes, 0, (int)len);
+        return new String(bytes, charset);
+    }
+
+    @ForceInline
+    private static void writeFast_short(MemorySegment segment, long offset, Charset charset, String string) {
+        byte[] bytes = string.getBytes(charset);
+        MemorySegment.copy(bytes, 0, segment, JAVA_BYTE, offset, bytes.length);
+        segment.set(JAVA_SHORT, offset + bytes.length, (short)0);
+    }
+
+    @ForceInline
+    private static int strlen_byte(MemorySegment segment, long start) {
+        // iterate until overflow (String can only hold a byte[], whose length can be expressed as an int)
```

Comment:
Perhaps we could see if functions such as `strnlen`, `wstrnlen` are supported, and, if so, create a trivial downcall method handle, and call that (as the underlying C function is likely to be a lot smarter than the simple loop here, and use vectorized instructions).